### PR TITLE
DAM Media modal and dashboard page

### PR DESCRIPTION
**Pull request by abaicus**
## All Submissions:

* [x] Have you followed the [Optimole Contributing guideline](https://github.com/Codeinwp/optimole-wp/blob/master/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change? 
 

## Changes proposed in this Pull Request:
- Changes the JS build process to split it into two separate flows for Dashboard and DAM - that's why so many files were moved;
- Removes the old media modal integration for cloud browsing and replaces it with the iframe that loads the dashboard and auto-logins the user;
- Adds a new dashboard page called `Optimole Assets`, which essentially loads a stripped-down version of the `Library` from the dashboard;
- Adds a REST route that inserts mock attachments using the remote URL. Attachments should not be imported twice if they already have been imported from DAM or offloaded from the website;
- If the attachment is sourced from the website but not offloaded, it will be reimported as a mock attachment because it's pretty hard to deduce which attachment we are using;

Closes Codeinwp/optimole-service#920;
Closes Codeinwp/optimole-service#919;

### Testing process for the plugin dashboard page (https://github.com/Codeinwp/optimole-service/issues/919):
- Activate the cloud browsing feature in the Optimole dashboard page;
- The link to the **Optimole Assets** should be visible in the dashboard sidebar after you save the settings;
- When deactivating this feature and saving, it should disappear;
- The filter for which sites you can browse will affect what is shown in the iframe (see **Known issues No.3 as you might stumble upon it**);
- We should check that everything works as expected on the dashboard page (i.e., all operations you can do on the dashboard should work fine there;
- Essentially, this loads the dashboard in an iframe, so if they work there, they should work here as well;

<details>
<summary>Screenshot of the dashboard iframe</summary>

image

</details>

---

### Testing for the Media Modal integration (https://github.com/codeinwp/optimole-service/issues/920):
- The Cloud browsing feature should be enabled;
- This will add an additional tab in the media modal in all contexts:
  - Adding an image into the editor, or replacing it - this allows you to select a single image;
  - Adding a gallery into the editor, or appending photos to an existing gallery - this allows a selection of maximum 10 images;
  - Setting a featured image to a post;
  - Essentially it should appear in all contexts where the media modal is opened to insert images;
- The new tab will allow you to select images straight from the dashboard and import them to your website;
- It will allow single selection for image selection, and multiple selection for galleries;
- When you import the images, some dummy attachments linking to the images in your optimole account will be created - these can be visible inside the **Media Library**;
- When attempting to insert duplicates that were previously imported there are two cases we treat and one we don't:
  - ✅ If the image was already imported through the DAM, it won't be imported again. Instead we will use the initial attachment we created;
  - ✅ If the site has images offloaded, it will use those attachments instead of adding new ones;
  - ❎ If the image is already sourced from the website but not offloaded, we import it again - we have no way of telling they are the same image for sure, unfortunately;
  
<details>
<summary>Screenshot of the Media Modal in gallery context</summary>

image

</details>

  
<details>
<summary>Screenshot of the Media Modal in single selection context</summary>

image

</details>

### Plugins I have tested with:
- Elementor - it seems that everything works as expected. If you find weird edge cases, do let me know;
- WooCommerce - works as expected with one exception (see Known Issues No. 1);
- Otter Blocks - mainly using blocks with images. Thins work as expected;

---

### Known issues:
1. In WooCommerce, when adding images to the product gallery, I couldn't find a way to allow the user to select multiple images;
2. In some instances, images are not resized, unless lazyloaded, or using an image size; 
3. When changing the cloud browsing accessible sites, if you just refresh the DAM dashboard page, it will still show the old sites. You have to manually navigate to that page again by clicking the dashboard link or focusing the browser addres bar and hitting `enter`. I don't know what exactly happens here, as things seem to be sent properly between the dashboard and the client website;
4. There is an edge case inside the customizer which I couldn't figure out. When editing widgets from there (e.g., Neve footer widgets) and attempting to add images to a gallery, the iframe simply disappears - I couldn't find a fix for this unfortunately. 

## Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

**I hope this information is enough, but please, do not hesitate to reach out and ask questions if you have any doubts.** 👍🏻

### CODE REVIEW

**Comment by abaicus** [comment]
@HardeepAsrani, thanks for the feedback! 

Yes, the changes are deployed. I am not sure what happened there. Is that happening with the dev build? It would help if maybe you could check this website as well: 

https://ed-peccary-moya.instawp.xyz

I will provide the credentials on Slack.

File: inc/dam.php [hunk 1]
```diff
@@ -0,0 +1,670 @@
...
+		$metadata = [];
+
+		$metadata['file']      = '/id:' . $image['meta']['resourceS3'] . '/' . get_home_url() . '/' . $filename;
+		$metadata['mime-type'] = $image['meta']['mimeType'];
+
+		if ( isset( $image['meta']['filesize'] ) ) {
+			$metadata['filesize'] = $image['meta']['fileSize'];
+		}
+
+		// todo: remove this after we deploy https://github.com/Codeinwp/optimole-service/pull/970
```
[the remark this answers is not in the document]
Comment (abaicus) [reply inc/dam.php]: @GrigoreMihai Thank you! I've removed it! 👍🏻

**Comment by rodica-andronache** [comment]
@abaicus great work, it looks really nice 🚀  Here are a few findings, I'm not sure if there is something I'm missing or not, but:

For https://github.com/Codeinwp/optimole-service/issues/919

- [ ] Deleting and creating folders, uploading images, moving images is not visible in the iframe, even if I refresh the page or manually navigate to that page again. I’ve seen the changes only after a few minutes.

- [ ] The filter for which sites you can browse will affect what is shown in the iframe -> I didn’t manage to do this no matter how many times I refreshed the page or manually navigate to it. It shows all images, not just the ones from the site I select.

https://www.loom.com/share/455ee9c35cec4ce6910a1c639054eda8

For https://github.com/Codeinwp/optimole-service/issues/920

- [ ] In a similar way, in the Optimole tab from the media modal, creating folders, uploading images is not visible, until a few minutes later, when I open another media modal in another context. 

- [ ] In the Brizy and Beaver Builder media modals, the Optimole tab is not added https://vertis.d.pr/i/toLMt4. If I understood well, we need to add the tab in all contexts, for all plugins.

Thank you!

**Comment by abaicus** [comment]
@rodica-andronache Thank you for the thorough testing! 

>  In the Brizy and Beaver Builder media modals, the Optimole tab is not added https://vertis.d.pr/i/toLMt4. If I understood well, we need to add the tab in all contexts, for all plugins.

We're not going to add it to these builders for now. It previously didn't work for anything besides the single image insertion in the editor, so I think we can consider this for a future enhancement. 👍🏻 

---

> The filter for which sites you can browse will affect what is shown in the iframe -> I didn’t manage to do this no matter how many times I refreshed the page or manually navigate to it. It shows all images, not just the ones from the site I select.

This should be working fine now. 

---

>  Deleting and creating folders, uploading images, moving images is not visible in the iframe, even if I refresh the page or manually navigate to that page again. I’ve seen the changes only after a few minutes.

>  In a similar way, in the Optimole tab from the media modal, creating folders, uploading images is not visible, until a few minutes later, when I open another media modal in another context.

These seem related to CloudFront caching. I'll keep investigating and let you know when I manage to fix it.

**Comment by abaicus** [comment]
@AndreeaCristinaRadacina Are you using the `dev` build from the comment above? Otherwise, this will not work. If yes, could you please share what the browser console is saying?

**Comment by abaicus** [comment]
@JohnPixle Thank you for the feedback on this. I will move it to another issue once this works fine, and tackle it if I still get the time before the code-freeze on Monday. 🚀

**Comment by abaicus** [comment]
@rodica-andronache Things should be in working order now. I managed to fix this. 🚀

**Comment by rodica-andronache** [comment]
@abaicus awesome, creating/deleting folders, uploading images works like a charm now ❤️ 
I still have issues with filtering the sites from which to show the images https://www.loom.com/share/f124b373bf504a63944e9089eb726d35

Also, this small thing, when you scroll at the end of the images this happens https://vertis.d.pr/v/pdrkor

**Comment by rodica-andronache** [comment]
@abaicus as we've talked on Slack, I've tested again the issue with site filtering and everything's working well. Indeed those images were directly uploaded, so that's why they were showing up in all cases. Thanks for explaining 🚀

**Comment by abaicus** [comment]
> Also, this small thing, when you scroll at the end of the images this happens https://vertis.d.pr/v/pdrkor

Regarding this: I sought a solution but couldn't come up with anything because of how the scroll behaves on this page and because we are in an iframe that cannot be higher than the current viewport - otherwise, you wouldn't be able to see the part out of the screen. 

We will leave it like this and address it in the future.

**Comment by irinelenache** [comment]
@abaicus Great job 🚀 Had a look and found a few issues:
- [ ] When i'm inside a nested folder and i move some images from there to another folder, it works but some console errors are thrown (the number of errors depends on the level of nesting) https://vertis.d.pr/v/UkfkEz
- [ ] When i'm using the bulk move feature, the sidebar toggle remains active and clickable (It works fine with the other pop-ups) https://vertis.d.pr/v/kHZBxs
- [ ] In Otter, if i'm creating a Section and use an image as background with a filename like `something (1).jpg`, it won't work in frontend because of the `)` at the end. If the image is uploaded through Wordpress directly, the character is striped and it works fine https://vertis.d.pr/v/uBup6F
- [ ] If i'm trying to offload some images, i'm not redirected to the Cloud integration page and the offload doesn't start https://vertis.d.pr/v/GBsmWL

The issues can be checked here:
```
URL: https://mole-dam.vertisite.cloud
Username: lisolusofe1875
Password: lParZuUDoMS1eXyW50fh
```

**Comment by abaicus** [comment]
@irinelenache thank you for pointing these out! 🚀  

>  In Otter, if i'm creating a Section and use an image as background with a filename like something (1).jpg, it won't work in frontend because of the ) at the end. If the image is uploaded through Wordpress directly, the character is striped and it works fine https://vertis.d.pr/v/uBup6F


This sounds like an unrelated issue. We should sanitize the image name when uploading files in the dashboard - there is where we allow parentheses in the file naming. I opened https://github.com/Codeinwp/optimole-service/issues/994 to handle this there.

I have fixed the offload redirects and visual issues you mentioned. 👍🏻 

Please let me know if you find any other issues! Thanks!

**Comment by irinelenache** [comment]
@abaicus The mentioned issues are fixed now, thank you ✅

**Pull request merged**